doc/man/lore.1
.TH GENERATELORE "1" "October 2002" "" ""
.SH NAME
lore \- convert documentations formats
.SH SYNOPSIS
.B lore [-l \fIlinkrel\fR] [-d \fIdocsdir\fR]  [-i \fIinput\fR] [-o \fIoutput\fR] [--config attribute[=value] [...]] [-p] [file [...]]
.PP
.B lore --help
.SH DESCRIPTION
.PP
The \fB\--help\fR prints out a usage message to standard output.
.TP
\fB-p\fR, \fB--plain\fR
Use non-flashy progress bar \- one file per line.
.TP
\fI-l\fR, \fI--linkrel\fR
Where non-document links should be relative to.
.TP
\fI-d\fR, \fI--docsdir\fR
Where to look for \fB.html\fR files if no files are given.
.TP
\fI-i\fR, \fI--input\fR
Input format. New input formats can be dynamically registered. Lore itself
comes with "lore" (the standard format), "mlore" (allows LaTeX equations)
and "man" (man page format).
.TP
\fI-o\fR, \fI--output\fR
Output format. Available output formats depend on the input. For the core
formats, lore and mlore support html, latex and lint, while man allows
lore.
.TP
\fI--config\fR
Add input/output-specific information. 
HTML output allows for 'ext=<extension>',
'template=<template>' and 'baseurl=<format string for API URLs>'. LaTeX
output allows for 'section' or 'chapter' in Lore, and nothing in Math-Lore. 
Lore output allows for 'ext=<extension>'. Lint output allows nothing.
Note that disallowed \fI--config\fR options are merely ignored, and do
not cause errors.
.SH DESCRIPTION
The exact HTML syntax to follow is documented in Twisted's doc-standard.html.
If no files are given, all *.html documents in docsdir will be processed.
.SH AUTHOR
Written by Moshe Zadka
.SH "REPORTING BUGS"
Report bugs to <twisted-python@twistedmatrix.com>.
.SH COPYRIGHT
Copyright \(co 2003 Matthew W. Lefkowitz
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


NEWS
Ticket numbers in this file can be looked up by visiting
http://twistedmatrix.com/trac/ticket/<number>

Twisted Core 16.0.0pre2 (2016-02-28)
====================================

Features
--------
 - twisted.cred.strports has been ported to Python 3. (#8216)

Deprecations and Removals
-------------------------
 - The __version__ attribute of former subprojects (conch, mail,
   names, news, pair, runner, web, and words) is deprecated in
   preference to the central twisted.__version__. (#8219)

Other
-----
 - #8220


Twisted Conch 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Mail 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre2 (2016-02-28)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre2 (2016-02-28)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre2 (2016-02-28)
===================================

No significant changes have been made for this release.


Twisted Words 16.0.0pre2 (2016-02-28)
=====================================

No significant changes have been made for this release.


Twisted Core 16.0.0pre1 (2016-02-16)
====================================

Features
--------
 - todo parameter for IReporter.addExpectedSuccess and
   IReporter.addUnexpectedSuccess is no longer required. If not
   provided, a sensible default will be used instead. (#4811)
 - A new string endpoint type, "tls:", allows for properly-verified
   TLS (unlike "ssl:", always matching hostname resolution with
   certificate hostname verification) with faster IPv4/IPv6
   connections.  This comes with an accompanying function,
   twisted.internet.endpoints.wrapClientTLS, which can wrap an
   arbitrary client endpoint with client TLS. (#5642)
 - twisted.python.filepath.makedirs accepts an ignoreExistingDirectory
   flag which ignore the OSError raised by os.makedirs if requested
   directory already exists. (#5704)
 - twisted.protocols.amp has been ported to Python 3. (#6833)
 - twisted.internet.ssl.trustRootFromCertificates returns an object
   suitable for use as trustRoot= to
   twisted.internet.ssl.optionsForClientTLS that trusts multiple
   certificates. (#7671)
 - twisted.python.roots is now ported to Python 3. (#8131)

Bugfixes
--------
 - Expected failures from standard library unittest no longer fail
   with Trial reporters. (#4811)
 - twisted.internet.endpoints.HostnameEndpoint.connect no longer fails
   with an AlreadyCalledError when the Deferred it returns is
   cancelled after all outgoing connection attempts have been made but
   none have yet succeeded or failed. (#8014)
 - twisted.internet.task.LoopingCall.withCount when run with internal
   of 0, now calls the countCallable with 1, regardless of the time
   passed between calls. (#8125)

Improved Documentation
----------------------
 - The Twisted Project has adopted the Contributor Covenant as its
   Code of Conduct. (#8173)

Deprecations and Removals
-------------------------
 - twisted.internet.task.LoopingCall.deferred is now deprecated. Use
   the deferred returned by twisted.internet.task.LoopingCall.start()
   (#8116)
 - twisted.internet.gtkreactor, the GTK+ 1 reactor deprecated since
   Twisted 10.1, has been removed. This does not affect the GTK2,
   GLib, GTK3, or GObject-Introspection reactors. (#8145)
 - twisted.protocols.mice, containing a Logitech MouseMan serial
   driver, has been deprecated. (#8148)

Other
-----
 - #6842, #6978, #7668, #7791, #7881, #7943, #7944, #8050, #8104,
   #8115, #8119, #8122, #8139, #8144, #8154, #8162, #8180, #8187


Twisted Conch 16.0.0pre1 (2016-02-16)
=====================================

Features
--------
 - twisted.conch now uses cryptography instead of PyCrypto for its
   underlying crypto operations. (#7413)
 - twisted.conch.ssh.keys is now ported to Python 3. (#7998)

Bugfixes
--------
 - twisted.conch.ssh.channel.SSHChannel's getPeer and getHost methods
   now return an object which provides IAddress instead of an old-
   style tuple address. (#5999)
 - twisted.conch.endpoint.SSHCommandClientEndpoint, when
   authentication is delegated to an SSH agent, no longer leaves the
   agent connection opened when connection to the server is lost.
   (#8138)

Other
-----
 - #7037, #7715, #8200, #8208


Twisted Mail 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Names 16.0.0pre1 (2016-02-16)
=====================================

No significant changes have been made for this release.


Twisted News 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Pair 16.0.0pre1 (2016-02-16)
====================================

No significant changes have been made for this release.


Twisted Runner 16.0.0pre1 (2016-02-16)
======================================

No significant changes have been made for this release.


Twisted Web 16.0.0pre1 (2016-02-16)
===================================

Features
--------
 - twisted.web.http_headers._DictHeaders now correctly handles
   updating via keyword arguments in Python 3 (therefore
   twisted.web.http_headers is now fully ported to Python 3). (#6082)
 - twisted.web.wsgi has been ported to Python 3. (#7993)
 - twisted.web.http_headers.Headers now accepts both Unicode and
   bytestring keys and values, encoding to iso-8859-1 and utf8
   respectively. (#8129)
 - twisted.web.vhost ported to Python 3. (#8132)

Bugfixes
--------

twisted/lore/scripts/lore.py
# Copyright (c) 2001-2008 Twisted Matrix Laboratories.
# See LICENSE for details.

"""
Implementation of the lore command line program.
"""

import sys

from zope.interface import Interface, Attribute

from twisted.lore import process, htmlbook
from twisted.lore.indexer import Indexer, TableOfContents
from twisted.lore.numberer import Numberer

from twisted.python import usage, plugin as oldplugin, reflect
from twisted import plugin as newplugin


class IProcessor(Interface):
    """
    A lore document transformation plugin.

    Processors provide generators for different output formats.
    """

    factory = Attribute(
        """
        An object with I{generate_}-prefixed methods for each output format
        supported.  For example, to customize HTML output, this object should
        define a method named C{generate_html}.
        """)



class Options(usage.Options):

    optFlags = [
        ["plain", 'p', "Report filenames without progress bar"],
        ["null", 'n', "Do not report filenames"],
        ["number", 'N', "Add chapter/section numbers to section headings"]]

    optParameters = [
        ["input", "i", 'lore'],
        ["inputext", "e", ".xhtml",
         "The extension that your Lore input files have"],
        ["docsdir", "d", None],
        ["linkrel", "l", ''],
        ["output", "o", 'html'],
        ["index", "x", None,
         "The base filename you want to give your index file"],
        ["book", "b", None, "The book file to generate a book from"],
        ["prefixurl", None, "",
         "The prefix to stick on to relative links; only useful when "
         "processing directories"]]

    #zsh_altArgDescr = {"foo":"use this description for foo instead"}
    #zsh_multiUse = ["foo", "bar"]
    #zsh_mutuallyExclusive = [("foo", "bar"), ("bar", "baz")]
    #zsh_actions = {"foo":'_files -g "*.foo"', "bar":"(one two three)"}
    #zsh_actionDescr = {"logfile":"log file name", "random":"random seed"}
    zsh_extras = ["*:files:_files"]


    def __init__(self, *args, **kw):
        usage.Options.__init__(self, *args, **kw)
        self.config = {}


    def opt_config(self, s):
        if '=' in s:
            k, v = s.split('=', 1)
            self.config[k] = v
        else:
            self.config[s] = 1


    def parseArgs(self, *files):
        self['files'] = files



def _getProcessorModule(input, output, config):
    """
    Find the module capable of processing the specified input type.

    @type input: C{str}
    @param input: A string describing the input format.  eg, C{'html'}.
    """
    plugins = oldplugin._getPlugIns("lore")
    for plug in plugins:
        if plug.tapname == input:
            return plug.load()
    else:
        plugins = newplugin.getPlugins(IProcessor)
        for plug in plugins:
            if plug.name == input:
                return reflect.namedModule(plug.moduleName)
        else:
            # try treating it as a module name
            try:
                return reflect.namedModule(input)
            except ImportError:
                print '%s: no such input: %s' % (sys.argv[0], input)
                return



def getProcessor(input, output, config):
    """
    Get the generate method of an output generator for the specified input and
    output formats.

    @return: A callable like L{HTMLGenerator.generate}
    """
    # XXX deprecate this
    module = _getProcessorModule(input, output, config)
    if module is None:
        return None
    try:
        return process.getProcessor(module, output, config)
    except process.NoProcessorError, e:
        print "%s: %s" % (sys.argv[0], e)



def getGenerator(input, output, config):
    """
    Get an output generator for the specified input and output formats.

    @return: An instance of a class like L{HTMLGenerator}.
    """
    module = _getProcessorModule(input, output, config)
    if module is None:
        return None
    return process.getOutputGenerator(module, output, config)



def getWalker(generator, opt):
    """
    Get a visitor appropriate for the specified options.
    """
    klass = process.Walker
    if opt['plain']:
        klass = process.PlainReportingWalker
    if opt['null']:
        klass = process.NullReportingWalker
    return klass(
        generator,
        opt['inputext'], opt['linkrel'],
        opt.config['template'])



def runGivenOptions(opt):
    """
    Do everything but parse the options; useful for testing.  Returns a
    descriptive string if there's an error.
    """
    book = None
    if opt['book']:
        book = htmlbook.Book(opt['book'])

    indexer = Indexer()
    toc = TableOfContents()

    generator = getGenerator(opt['input'], opt['output'], opt.config)
    if generator is None:
        # XXX Not an awesome return value
        return 'getProcessor() failed'

    walker = getWalker(generator, opt)

    if opt['files']:
        for filename in opt['files']:
            walker.walked.append(('', filename))
    elif book is not None:
        for filename in book.getFiles():
            walker.walked.append(('', filename))
    else:
        walker.walkdir(opt['docsdir'] or '.', opt['prefixurl'])

    if opt['index']:
        indexFilename = opt['index']
    elif book is not None:
        indexFilename = book.getIndexFilename()
    else:
        indexFilename = None

    if indexFilename:
        indexer.setIndexFilename("%s.%s" % (indexFilename, opt['output']))
    else:
        indexer.setIndexFilename(None)

    ## TODO: get numberSections from book, if any
    numberer = Numberer()
    numberer.numberSections = opt['number']

    walker.generate(book, indexer, toc)

    if walker.failures:
        for (file, errors) in walker.failures:
            for error in errors:
                print "%s:%s" % (file, error)
        return 'Walker failures'



def run():
    opt = Options()
    try:
        opt.parseOptions()
    except usage.UsageError, errortext:
        print '%s: %s' % (sys.argv[0], errortext)
        print '%s: Try --help for usage details.' % sys.argv[0]
        sys.exit(1)

    result = runGivenOptions(opt)
    if result:
        print result
        sys.exit(1)


if __name__ == '__main__':
    run()

